library/src/main/java/com/google/android/exoplayer/extractor/ts/MpaReader.java

```diff
@@ -0,0 +1,293 @@
+     *
+     * @param pesBuffer The buffer whose position should be advanced.
+     * @return True if a sync word position was found. False otherwise.
+     */
+    private boolean skipToNextSync(ParsableByteArray pesBuffer) {
+        byte[] mpaData = pesBuffer.data;
+        int startOffset = pesBuffer.getPosition();
+        int endOffset = pesBuffer.limit();
+        for (int i = startOffset; i < endOffset - 1; i++) {
+            int syncBits = ((mpaData[i] & 0xFF) << 8 ) | (mpaData[i + 1] & 0xFF);
+            if ((syncBits & 0xFFF0) == 0xFFF0) {
+                hasCrc = (mpaData[i + 1] & 0x1) == 0;
+                pesBuffer.setPosition(i);
+                return true;
+            }
+        }
+        pesBuffer.setPosition(endOffset);
+        return false;
+    }
+
+    /**
+     * Calculates MPEG Audio frame size
+     *
+     * @param layer The MPEG layer
+     * @param LSF Low Sample rate Format (MPEG 2)
+     * @param bitrate The bitrate in bits per second
+     * @param samplesPerSec The sampling rate in hertz
+     * @param -paddingSize
+     * @return Frame size in bytes
+     */
+    private static int CalcMpaFrameSize (int layer, int LSF, int bitrate, int samplesPerSec, int paddingSize) {
+        return (int)(Math.floor(MPA_COEFFICIENTS[LSF][layer] * bitrate / samplesPerSec) + paddingSize) * MPA_SLOT_SIZE[layer];
+    }
+
+    /**
+     * Parses the sample header.
+     */
+    private void parseHeader() {
+        int headerLength = hasCrc ? HEADER_SIZE + CRC_SIZE : HEADER_SIZE;
+
+        if (!hasOutputFormat) {
+            mpaScratch.setPosition(0);
+            mpaScratch.skipBits(12);
+            int isLSF = (!mpaScratch.readBit()) ? 1 : 0;
+            int layer = mpaScratch.readBits(2) ^ 3;
+            mpaScratch.skipBits(1);
+            int audioObjectType = 32 + layer;
+            int bitRate = MPA_BITRATES[isLSF][layer][mpaScratch.readBits(4)];
+            int sampleRate = MPA_SAMPLING_RATES[3 - isLSF][mpaScratch.readBits(2)];
+            int sampleRateIndex = CodecSpecificDataUtil.getSampleRateIndex(sampleRate);
+            int paddingBit = (mpaScratch.readBit()) ? 1 : 0;
+            mpaScratch.skipBits(1);
+            int channelConfig = mpaScratch.readBits(2) == 3 ? 1 : 2;
+
+            byte[] audioSpecificConfig = CodecSpecificDataUtil.buildAudioSpecificConfig(
+                    audioObjectType, sampleRateIndex, channelConfig);
+            Pair<Integer, Integer> audioParams = CodecSpecificDataUtil.parseAudioSpecificConfig(
+                    audioSpecificConfig);
+
+            // need to investigate how to detect if the mpeg decoder supports Layers other than Layer III
+            MediaFormat mediaFormat = MediaFormat.createAudioFormat(/*isLSF == 1 ?*/ MimeTypes.AUDIO_MPEG/* : MimeTypes.AUDIO_MP1L2*/,
+                    MediaFormat.NO_VALUE, audioParams.second, audioParams.first,
+                    Collections.singletonList(audioSpecificConfig));
```

Comment:
i have a bug in the buildAudioSpecificConfig, the channel config is 4 bits, could you try the following patch?

```java
diff --git a/library/src/main/java/com/google/android/exoplayer/util/CodecSpecificDataUtil.java b/library/src/main/java/com/google/android/exoplayer/util/CodecSpecificDataUtil.java
index 8e9bc9b..26094a0 100644
--- a/library/src/main/java/com/google/android/exoplayer/util/CodecSpecificDataUtil.java
+++ b/library/src/main/java/com/google/android/exoplayer/util/CodecSpecificDataUtil.java
@@ -77,8 +77,8 @@ public final class CodecSpecificDataUtil {
       int frequencyIndex = (audioSpecificConfig[1] & 0x1E) >> 1;
       Assertions.checkState(frequencyIndex < 13);
       int sampleRate = AUDIO_SPECIFIC_CONFIG_SAMPLING_RATE_TABLE[frequencyIndex];
-      int channelCount = (audioSpecificConfig[1] & 0x01) << 2 |
-              ((audioSpecificConfig[2] >> 6) & 0x03);
+      int channelCount = (audioSpecificConfig[1] & 0x01) << 3 |
+              ((audioSpecificConfig[2] >> 5) & 0x07);
       return Pair.create(sampleRate, channelCount);
     }
   }
@@ -108,8 +108,8 @@ public final class CodecSpecificDataUtil {
               (audioObjectTypeExt >> 3) & 0x07);
       audioSpecificConfig[1] = (byte) ((audioObjectTypeExt << 5) & 0xE0 |
               (sampleRateIndex    << 1) & 0x1E |
-              (channelConfig      >> 2) & 0x01);
-      audioSpecificConfig[2] = (byte) ((channelConfig << 6) & 0xC0);
+              (channelConfig      >> 3) & 0x01);
+      audioSpecificConfig[2] = (byte) ((channelConfig << 5) & 0xE0);
       return audioSpecificConfig;
     }
   }
```